### ISSUE TITLE
Characters from Zapf Dingbats in PDF are not rendered (poppler/fontconfig)

### ISSUE THREAD

**Issue by Mathnerd314** [NixOS/nixpkgs#90705]
Opening [this PDF](https://web.archive.org/web/20160818132612if_/http://thyer.name/phd-thesis/thesis-thyer.pdf), it's missing the check / X marks ✓ ✗ on pages 84-87. Chrome, Firefox, Adobe Reader, and mupdf all render fine, but Evince draws ▯ boxes and Okular draws nothing, so it seems to be poppler-related.

Investigating the properties in Okular/evince it seems that the glyphs are from ZapfDingbats, and it's trying to substitute with Noto Sans Regular. According to https://gitlab.freedesktop.org/poppler/poppler/-/issues/926 the issue is "a broken system fontconfig configuration". I haven't done anything noteworthy with the fonts so the issue is really with NixOS's fontconfug defaults.

In particular NixOS doesn't ship the URW fonts recommended by poppler; AFAICT they are not even packaged. Instead there is the default font FreeSerif. Maybe the Ghostscript fonts would fix it too but I haven't tested.

**To Reproduce**
1. Open https://web.archive.org/web/20160818132612if_/http://thyer.name/phd-thesis/thesis-thyer.pdf in evince/okular
2. Go to page 84

**Expected behavior**
Checkmarks render correctly

**Screenshots**
Correct rendering:
![Correct](https://user-images.githubusercontent.com/322214/84972485-31061f80-b0dc-11ea-9783-f9498d60141c.png)

**Fix**
FreeSerif contains the characters and seems to render them just fine. I added this to local.conf:
```
  <alias binding="same">
    <family>Zapf Dingbats</family>
    <accept>
      <family>FreeSerif</family>
    </accept>
  </alias>
```
But there are [14 standard fonts](https://en.wikipedia.org/wiki/PDF#Standard_Type_1_Fonts_(Standard_14_Fonts)), so this needs more work.

**Notify maintainers**
Nobody specifically maintains the default fonts in NixOS. The last interesting commits were https://github.com/NixOS/nixpkgs/pull/67667 by @worldofpeace / @jtojnar and https://github.com/NixOS/nixpkgs/commit/2214b638a7ddc7845eacfdc74fedd786c7d6e344 by @ttuegel.

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.6.16, NixOS, 20.09 (Nightingale)` (based on 029a5de08390bb03c3f44230b064fd1850c6658a)
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`

Maintainer information:
```yaml
# a list of nixos modules affected by the problem
module: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/fonts/fontconfig.nix
```

**Comment by jtojnar** [comment]
Seems to work for me with DejaVu Sans (installed through `fonts.enableDefaultFonts = true` NixOS option and likely matched by being set as default sans serif font through `fonts.fontconfig.defaultFonts.sansSerif` option):

![image](https://user-images.githubusercontent.com/705123/84989541-1ff0f900-b144-11ea-8ad7-dad3f78ae56d.png)

Also Type1 fonts are disabled by default, see `fonts.fontconfig.allowType1` option.

**Comment by Mathnerd314** [comment]
Right, so I'm running KDE, which sets the default fonts to Noto / Hack: https://github.com/NixOS/nixpkgs/blob/a5a52e8c73ed2bb52b5fcb0bc93a77769412e72e/nixos/modules/services/x11/desktop-managers/plasma5.nix#L322-L327

DejaVu is available, because x11 is enabled: https://github.com/NixOS/nixpkgs/blob/a5a52e8c73ed2bb52b5fcb0bc93a77769412e72e/nixos/modules/services/x11/xserver.nix#L811

But it isn't what fontconfig is choosing.

**Comment by Mathnerd314** [comment]
So one option might be to set all the Noto Sans fonts as fallbacks to Noto Sans, similar to what this fontconfig does:
https://github.com/nuew/rc/blob/ddc70c6a00a9b008bccd0fd1b06d1876c04405cc/.config/fontconfig/fonts.conf#L46

**Comment by Mathnerd314** [comment]
I looked at this again and apparently only 2 fonts are really needed, Symbol and Zapf Dingbats. https://freedesktop.org/wiki/Software/poppler/#suggestedfontconfigfontsubstitutions

So we have 2 options:

1. Package the URW fonts (there are [fontconfig aliases](https://github.com/ArtifexSoftware/urw-base35-fonts/tree/master/fontconfig) which do as poppler recommends, Standard Symbols PS / D050000L) 
2. Alias the fonts to some other font that has the Unicode characters, e.g. DejaVu Sans

**Comment by jtojnar** [comment]
I am worried about the possible breakage so it should probably be opt in.

I like the URW variant slightly more as they seem to have been explicitly designed to replace the fonts mandated by the PostScript standard, whereas it is not clear to me how compatible DejaVu Sans is.

**Comment by Mathnerd314** [comment]
Alright, what about a `fonts.enablePdfFonts` switch that adds the URW fonts and corresponding fontconfig? I guess it should be distinct from `fonts.enableDefaultFonts` but similarly enabled by default with X11.

**Comment by jtojnar** [comment]
Not sure about default but that sounds good to me.

**Comment by Mathnerd314** [comment]
Well, the right default is to set it if a poppler-using application *and* X11 are included in the system, but I don't think there's an easy way to scan `environment.systemPackages` for poppler.

**Comment by Mathnerd314** [comment]
There is already [enableGhostscriptFonts](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/fonts/ghostscript.nix), maybe setting that is enough.

**Comment by TLATER** [comment]
> I am worried about the possible breakage so it should probably be opt in.

What breakage are you worried about specifically? These settings may fall back to fonts unintended by the user, but when using the fallback config this would specifically apply to the replacement fonts (see e.g. [urw-fallback-generics](https://github.com/ArtifexSoftware/urw-base35-fonts/blob/master/fontconfig/urw-fallback-generics.conf)) which aren't packaged on NixOS anyway.

User config still always has priority, so the only case in which this breaks is if someone packaged and installed those non-free fonts downstream. I think this is a very unlikely use case, and publishing that backwards incompatibility as part of the release notes seems like enough precaution; the upside will be to fix rendering for all the people who don't care to delve deeply into font configuration.

**Comment by jtojnar** [comment]
Sorry, I do not remember any more, maybe I misread the configs. Though fontconfig configuration is always risky. But If someone wants to package the fonts and open a PR for the option, I agree that having it in release notes should be sufficient precaution.
